### PR TITLE
fix(audio): drain Opus lookahead on finish

### LINKED ISSUE #2994: moq-audio: finish drops the Opus encoder's lookahead tail
`encode::Producer::finish` drains the encoder's sample buffer (and, since #2992, the resampler's), but nothing drains the codec's own lookahead. Opus runs ~312 frames (6.5 ms at 48 kHz) of encoder delay: the last 312 input samples fed to the encoder have not been emitted in any packet until later input pushes them through.

`finish` encodes every full pending frame and zero-pads one final partial frame. That padding flushes the lookahead only by accident, when the partial frame happens to contain at least 312 samples of padding:

- If the pending samples land exactly on an encoder-frame boundary, no padded packet is produced at all and the last 312 source frames never leave the encoder.
- If the final partial frame leaves fewer than 312 padding samples, the same number of real tail samples stay behind.

This predates #2992 and is the codec-layer analogue of the resampler bugs fixed there: a stage with internal delay whose terminal drain does not account for it. The existing `finish` tests count packets rather than decoding content, which is what lets it hide (the packet-count test's final frame carries ~940 padding samples, comfortably past the lookahead).

Fix shape: a codec-aware terminal drain on `Encoder`, feeding enough silence to push the lookahead out without extending the logical track, mirroring how `Resampler::flush` repays `output_delay`. Regression test: publish a signal with an impulse in the last 312 samples, decode the published packets, and assert the impulse survives, including the exact-frame-boundary case.

Found by Codex adversarial review on #2992; split out because it needs encoder surface work and affects the non-resampled path too.

(Written by Fable 5)

### PR DESCRIPTION
**Pull request by kixelated**
## Summary

- drain libopus lookahead when an audio producer finishes, including tracks ending exactly on an encoder-frame boundary
- return terminal discard-padding metadata and publish the logical source endpoint before terminal codec packets
- expose Legacy endpoint markers as ordered metadata in the shared Rust and TypeScript Hang consumers
- signal fresh codec epochs with empty groups and reset native and browser decoders before resumed media
- preserve empty-group codec boundaries when latency skipping crosses them, including delayed group FINs
- account for Opus pre-skip when mapping browser decoder timestamps, then trim decoded output to the source endpoint
- keep browser decoder warm-up limited to initial playback setup so reset epochs retain their source audio
- reset libopus state with each producer epoch so abandoned lookahead cannot leak into a later source or terminal drain

The root cause was that `Producer::finish` drained only producer and resampler buffers. Libopus retains 312 frames of codec lookahead, while decoding the packets needed to release that tail also produces zero-padding that must be discarded. `reset_epoch` likewise cleared outer buffers but left encoder and receiver codec history live. Browser playback also compared endpoint timestamps against decoder output without removing the Opus pre-skip delay. Finally, latency skipping could discard a timestamp-less group before its delayed FIN established that it was a codec boundary. Browser decoder resets must preserve the one-time playback warm-up state because their next callbacks contain resumed source audio.

Closes #2994.

## Public API changes

- adds `moq_audio::encode::Encoder::finish(self, pcm: &[f32]) -> Result<Finish, Error>`, an additive terminal operation for partial-frame padding and codec drain
- adds `moq_audio::encode::Finish` with packet accessors and `discard_padding()` in decoded frames per channel
- adds `moq_audio::decode::Decoder::reset()` for starting a fresh codec epoch
- adds `moq_mux::container::Consumer::end()` for the exclusive endpoint from the latest consumed container marker
- adds optional endpoint recognition to the Rust and TypeScript container format interfaces
- adds `Opus.preSkip` to `@moq/hang/util` for reading the codec delay from Opus configuration

## Test plan

- `just fix`
- `just check`
- `just test` (all affected JS packages and 1,257 Rust tests passed)
- `just drafts check`
- `just test smoke-full` (all 21 Rust, Python, JS, C, and GStreamer publisher/subscriber combinations passed)
- verified exact-frame and partial-frame Opus output preserves the tail and decodes to the original source length
- verified browser pre-skip handling keeps exact-frame source audio and drops only decoded drain output at the endpoint
- verified reset after a full frame emits no abandoned drain packet and no pre-reset impulse in the next epoch
- verified continued native subscriptions reset decoder state and reapply pre-skip after a discontinuity
- verified browser decoder warm-up drops only initial setup callbacks and does not discard resumed-epoch output
- verified Rust and TypeScript latency skipping preserves completed empty-group boundaries and waits for delayed FINs
- verified endpoint markers do not consume the first-media keyframe slot
- verified a rewound endpoint survives its discontinuity reset and trims the following drain packet

## Cross-package sync

The Legacy wire semantics are documented in `draft-lcurley-moq-hang.md`, including empty groups as codec epoch boundaries. Matching endpoint and discontinuity handling is implemented in `moq-mux`, `js/hang`, native `moq-audio`, and browser `js/watch`. Marker detection stays container-specific, so custom formats may still carry legitimate empty media payloads. FFI and command-line surfaces are unaffected.

(Written by GPT-5)

### CODE REVIEW

File: rs/moq-audio/src/encode/producer.rs [hunk 1]
```diff
@@ -331,6 +335,61 @@ impl<E: CatalogExt> Drop for Rendition<E> {
 mod tests {
 	use super::*;
 	use crate::Format;
+	use crate::decode::Decoder;
+
+	/// Opus retains its final lookahead until later input pushes it into a
+	/// packet. Ending exactly on a codec-frame boundary used to leave no pending
+	/// producer buffer to pad, so the track dropped that tail outright. A nearly
+	/// full partial frame failed the same way when its padding was too short.
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-audio/src/encode/producer.rs]: Updated the test comment to state the present-tense invariant for exact-frame and partial-frame terminal lookahead. (Written by GPT-5)

File: rs/moq-audio/src/encode/producer.rs [hunk 2]
```diff
@@ -287,18 +291,18 @@ impl<E: CatalogExt> Producer<E> {
...
-		let frame_samples = self.encoder.frame_size() * self.encoder.codec_channels() as usize;
-		if !self.pending.is_empty() {
-			self.pending.resize(frame_samples, 0.0);
-			let chunk = std::mem::take(&mut self.pending);
-			let packet = self.encoder.encode(&chunk)?;
-			let timestamp = self.timestamp(epoch_us)?;
-			self.publish(packet, timestamp)?;
+		let frame_size = self.encoder.frame_size();
+		let codec_rate = self.encoder.codec_rate();
+		let packets = self.encoder.finish(&self.pending)?;
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-audio/src/encode/producer.rs]: Addressed in 3a26c8d. reset_epoch now resets libopus state and clears the encoder's started state in addition to the producer and resampler buffers. Regressions verify that resetting after a full frame emits no abandoned terminal drain and that a pre-reset impulse does not appear in the next epoch. (Written by GPT-5)

File: rs/moq-audio/src/encode/producer.rs [hunk 3]
```diff
@@ -287,18 +291,18 @@ impl<E: CatalogExt> Producer<E> {
...
+		for packet in packets {
+			let timestamp = Self::timestamp(epoch_us, self.frames_produced, codec_rate)?;
+			Self::publish(&mut self.track, packet, timestamp)?;
+			self.frames_produced += frame_size as u64;
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-audio/src/encode/producer.rs]: Addressed in 3a26c8d. Encoder::finish now returns a Finish value with discard_padding metadata. The producer carries the logical endpoint in the existing empty Legacy marker, and the native consumer preserves that marker long enough to truncate decoded terminal output. The regression now asserts that both 960-frame and 860-frame inputs decode to exactly their original source length while retaining the tail impulse. (Written by GPT-5)

File: rs/moq-audio/src/encode/producer.rs [hunk 4]
```diff
@@ -258,10 +271,37 @@ impl<E: CatalogExt> Producer<E> {
...
+	/// Publish terminal packets after an empty frame that carries their logical endpoint.
+	fn publish_terminal(
+		track: &mut moq_mux::container::Producer<moq_mux::container::legacy::Wire>,
+		terminal: Terminal,
+	) -> Result<(), Error> {
+		track.write(MuxFrame {
+			timestamp: terminal.end,
+			payload: Bytes::new(),
+			keyframe: true,
+			duration: None,
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-audio/src/encode/producer.rs:311]: Addressed in 28a3f60a3. The shared Rust and TypeScript Legacy container paths now recognize empty-payload frames as ordered endpoint metadata while leaving custom formats free to carry legitimate empty media. Native moq-audio and browser js/watch both consume that endpoint and trim decoded terminal output, and the Hang draft now specifies the marker, terminal packet ordering, and exclusive trim rule. Regressions cover endpoint ordering, custom empty payloads, partial browser frames, and drain packets, and the full cross-language smoke matrix passes. (Written by GPT-5)

File: js/watch/src/audio/decoder.ts [hunk 5]
```diff
@@ -332,6 +337,10 @@ export class Decoder {
 			for (;;) {
 				const next = await consumer.next();
 				if (!next) break;
+				if (next.end !== undefined) {
+					this.#terminalEnd = next.end;
+					continue;
+				}
 
 				// Publisher rewound the timeline: flush + re-anchor before decoding the new frame.
 				this.#onDiscontinuity(next.discontinuity);
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/watch/src/audio/decoder.ts]: Fixed in a6385931e. The decoder now applies each ordered consumer result through Terminal.update before branching on endpoint metadata. A discontinuity clears the prior endpoint first, then preserves an endpoint carried by that same result, so the following drain packet is trimmed against the new epoch. Regressions cover the state transition and a stalled-group rewind where the marker and first terminal packet share the new discontinuity count. (Written by GPT-5)

File: js/hang/src/container/consumer.ts [hunk 6]
```diff
@@ -549,6 +557,17 @@ export class Consumer {
...
+					if (end !== undefined) {
+						this.#updateBuffered();
+						return {
+							frame: undefined,
+							group: seq,
+							discontinuity: this.#rewind.discontinuity,
+							continuous,
+							end,
+						};
+					}
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/hang/src/container/consumer.ts:578]: Fixed in a6385931e. Both the Rust and TypeScript container consumers now detect endpoint markers before advancing the media-frame index, so a leading marker does not consume the protocol-derived first-keyframe slot. Added Rust and TypeScript regressions that assert the first following media frame is still a keyframe. (Written by GPT-5)

File: js/watch/src/audio/decoder.ts [hunk 7]
```diff
@@ -332,6 +337,10 @@ export class Decoder {
 			for (;;) {
 				const next = await consumer.next();
 				if (!next) break;
+				if (next.end !== undefined) {
+					this.#terminalEnd = next.end;
+					continue;
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/watch/src/audio/decoder.ts:346]: Fixed in a6385931e. Decoder metadata is applied through Terminal.update before branching on the endpoint. A discontinuity clears the prior endpoint first and then stores an endpoint carried by the same ordered result, so its following drain remains trimmed. The stalled-group rewind regression covers this exact ordering. (Written by GPT-5)

File: js/watch/src/audio/decoder.ts [hunk 8]
```diff
@@ -448,6 +451,11 @@ export class Decoder {
 	#emit(sample: AudioData) {
 		const timestamp = sample.timestamp as Time.Micro;
 		const timestampMilli = Time.Milli.fromMicro(timestamp);
+		const frames = terminalFrames(sample, this.#terminal.end);
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/watch/src/audio/decoder.ts]: Fixed in 4e1588626. Browser playback now parses Opus pre-skip from OpusHead or dOps, removes that prefix once per codec epoch, and maps each decoded packet timestamp back by the codec delay before applying the exclusive endpoint. The exact-frame regression now keeps 648 source frames from the first packet and 312 source frames from the drain packet, for the original 960-frame length, instead of discarding the real tail. A discontinuity reapplies pre-skip for the resumed epoch. (Written by GPT-5)

File: rs/moq-audio/src/encode/producer.rs [hunk 9]
```diff
@@ -188,6 +196,7 @@ impl<E: CatalogExt> Producer<E> {
 		self.epoch_us = None;
 		self.frames_produced = 0;
 		self.pending.clear();
+		self.encoder.reset();
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-audio/src/encode/producer.rs:213]: Fixed in 4e1588626. Producer::reset_epoch now schedules an ordered empty-group boundary before the next packet, including forward timestamp gaps, while an explicit audio discontinuity resets the encoder immediately. The shared Rust and TypeScript container consumers advance their discontinuity counters at that boundary. Native consumers reset libopus, AAC, resampler, and terminal state, and browser playback resets and reconfigures AudioDecoder before resumed packets. The Hang draft specifies the empty-group codec boundary. An end-to-end regression verifies a continued subscription reapplies the 312-frame Opus pre-skip after reset and preserves the resumed source length. (Written by GPT-5)

File: js/watch/src/audio/decoder.ts [hunk 10]
```diff
@@ -324,17 +328,22 @@ export class Decoder {
...
 			for (;;) {
 				const next = await consumer.next();
 				if (!next) break;
-
-				// Publisher rewound the timeline: flush + re-anchor before decoding the new frame.
-				this.#onDiscontinuity(next.discontinuity);
+				if (this.#onNext(next)) {
+					decoder.reset();
+					decoder.configure(decoderConfig);
+				}
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/watch/src/audio/decoder.ts:344]: Fixed in 8a89f7836. Browser decoder warm-up now uses epoch-aware state and restarts it before decoder reset and reconfiguration. The regression confirms that exactly the first three callbacks are dropped both initially and after a reset. (Written by GPT-5)

File: rs/moq-mux/src/container/consumer.rs [hunk 11]
```diff
@@ -248,8 +263,15 @@ impl<F: Container> Consumer<F> {
 					}
 					// Cleanly finished group: advance to the next sequence.
 					Poll::Ready(Ok(None)) => {
+						let empty = group.empty;
 						self.pending.pop_front();
 						self.current += 1;
+						if empty {
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-mux/src/container/consumer.rs:269]: Fixed in 8a89f7836. Both shared Rust and TypeScript container consumers now detect cleanly finished empty groups in ranges discarded by latency skipping and propagate their discontinuity counters before resumed media. Rust and TypeScript regressions cover a missing or stalled sequence followed by an empty boundary and resumed frames. (Written by GPT-5)

File: js/hang/src/container/consumer.ts [hunk 12]
```diff
@@ -362,6 +366,7 @@ export class Consumer {
 				`skipping slow group: track=${this.#track.name} ${first.consumer.sequence} -> ${this.#active}`,
 			);
 
+			if (first.empty && first.consumer.done) this.#markDiscontinuity();
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/hang/src/container/consumer.ts]: Fixed in 4c91f3d50. The TypeScript consumer now promotes a completed empty head across a sequence gap even when it has no timestamp, surfaces its discontinuity, and only then delivers newer media. A regression uses a single later frame so no timestamp span can trigger the latency path. (Written by GPT-5)

File: rs/moq-mux/src/container/consumer.rs [hunk 13]
```diff
@@ -318,7 +337,13 @@ impl<F: Container> Consumer<F> {
 			if let Some((new_idx, _)) = next_group
 				&& should_skip
 			{
+				let discontinuities = self
+					.pending
+					.iter_mut()
+					.take(new_idx)
+					.fold(0, |count, group| count + u64::from(group.poll_empty(waiter)));
```
[the remark this answers is not in the document]
Comment (kixelated) [reply rs/moq-mux/src/container/consumer.rs]: Fixed in 4c91f3d50. Rust and TypeScript now keep any zero-frame group in order until its terminal state resolves, so latency skipping cannot outrun a delayed FIN. The regressions hold the marker open while newer media is buffered, verify that reading remains pending, then close it and verify the discontinuity is observed before resumed media. (Written by GPT-5)

**Comment by kixelated** [comment]
@codex review

Please review the current head, including the final named warm-up callback count requested in the review body.

(Written by GPT-5)

File: js/watch/src/audio/decoder.ts [hunk 14]
```diff
@@ -324,17 +329,23 @@ export class Decoder {
...
+			decoder.configure(decoderConfig);
 
 			for (;;) {
 				const next = await consumer.next();
 				if (!next) break;
-
-				// Publisher rewound the timeline: flush + re-anchor before decoding the new frame.
-				this.#onDiscontinuity(next.discontinuity);
+				if (this.#onNext(next)) {
+					warmup.reset();
```
[the remark this answers is not in the document]
Comment (kixelated) [reply js/watch/src/audio/decoder.ts]: Fixed in dbdb4289b. The three-callback drop is now limited to initial playback setup and is not restarted when AudioDecoder resets for a codec epoch. Warmup no longer exposes a reset operation, and the regression verifies that it stops dropping permanently after the configured initial callbacks. Opus pre-skip is still reapplied separately by Terminal for each epoch. (Written by GPT-5)

**Comment by kixelated** [comment]
@codex review

Please re-review the exact current head after the decoder warm-up reset fix.

(Written by GPT-5)
